src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicListener.cs

```diff
@@ -101,120 +110,98 @@ private void Dispose(bool disposing)
-                ref address),
-                "Failed to start listener.");
+            QuicExceptionHelpers.ThrowIfFailed(status, "ListenerStart failed.");
 
-            SetListenPort();
+            SOCKADDR_INET inetAddress = MsQuicParameterHelpers.GetINetParam(MsQuicApi.Api, _state.Handle, QUIC_PARAM_LEVEL.LISTENER, (uint)QUIC_PARAM_LISTENER.LOCAL_ADDRESS);
+            _listenEndPoint = MsQuicAddressHelpers.INetToIPEndPoint(ref inetAddress);
         }
 
         internal override void Close()
         {
             ThrowIfDisposed();
-
-            MsQuicApi.Api.ListenerStopDelegate(_ptr);
+            MsQuicApi.Api.ListenerStopDelegate(_state.Handle);
         }
 
-        private void SetListenPort()
+        private void StopAcceptingConnections()
         {
-            SOCKADDR_INET inetAddress = MsQuicParameterHelpers.GetINetParam(MsQuicApi.Api, _ptr, (uint)QUIC_PARAM_LEVEL.LISTENER, (uint)QUIC_PARAM_LISTENER.LOCAL_ADDRESS);
-
-            _listenEndPoint = MsQuicAddressHelpers.INetToIPEndPoint(ref inetAddress);
+            // TODO finalizers are called even if the object construction fails.
+            if (_state != null)
+            {
+                _state.AcceptConnectionQueue.Writer.TryComplete();
+            }
         }
 
-        internal unsafe uint ListenerCallbackHandler(ref ListenerEvent evt)
+        private static unsafe uint NativeCallbackHandler(
+            IntPtr listener,
+            IntPtr context,
+            ref ListenerEvent evt)
         {
-            try
+            if (evt.Type != QUIC_LISTENER_EVENT.NEW_CONNECTION)
             {
-                switch (evt.Type)
-                {
-                    case QUIC_LISTENER_EVENT.NEW_CONNECTION:
-                        {
-                            ref NewConnectionInfo connectionInfo = ref *(NewConnectionInfo*)evt.Data.NewConnection.Info;
-
-                            IPEndPoint localEndPoint = MsQuicAddressHelpers.INetToIPEndPoint(ref *(SOCKADDR_INET*)connectionInfo.LocalAddress);
-                            IPEndPoint remoteEndPoint = MsQuicAddressHelpers.INetToIPEndPoint(ref *(SOCKADDR_INET*)connectionInfo.RemoteAddress);
-
-                            MsQuicConnection msQuicConnection = new MsQuicConnection(localEndPoint, remoteEndPoint, evt.Data.NewConnection.Connection, _options.IdleTimeout);
-                            msQuicConnection.SetNegotiatedAlpn(connectionInfo.NegotiatedAlpn, connectionInfo.NegotiatedAlpnLength);
-
-                            _acceptConnectionQueue.Writer.TryWrite(msQuicConnection);
-                        }
-                        // Always pend the new connection to wait for the security config to be resolved
-                        // TODO this doesn't need to be async always
-                        return MsQuicStatusCodes.Pending;
-                    default:
-                        return MsQuicStatusCodes.InternalError;
-                }
+                return MsQuicStatusCodes.InternalError;
```

Comment:
What is the plan for logging inside System.Net.Quic?

For example, here there should be a log message that an unknown callback was received, along with the `evt.Type` value. A generic `InternalError` status code isn't enough to understand what went wrong.

(logging can be added in the future, not blocking for this PR)

Reply:
Found this - https://github.com/dotnet/runtime/issues/32072